README
Tie-LevelDB - A Perl Interface to the LevelDB database
======================================================

This module is based on LevelDB NoSQL database from Google.

  http://code.google.com/p/leveldb/

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

Module has no dependencies. LevelDB sources are bundled with this package.

COPYRIGHT AND LICENCE

Copyright (C) 2011 by Sokordia, s.r.o. http://www.sokordia.cz

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Tie::LevelDB.

0.01  Fri Jul 29 23:08:09 2011
	- original version; created by h2xs 1.23 with options
         --use-new-tests -A Tie::LevelDB
    - limited functionality: just Open, Put, Get and Delete

0.02  Sat Jul 30 09:21:19 2011
    - DB, Iterator, WriteBatch API covered
    - Perl tie (TIEHASH) support

0.03  Sat Jul 30 10:21:19 2011
    - Instructions for Snappy support
    - Warning about missing "port/" and "port/win/" suppressed
    - Storing undef values deletes the key automatically

0.04  Sat Jul 31 07:21:19 2011
    - Rewritted to support binary keys and values
    - Module passes stress-testing

0.05  Mon Sep 18 14:33:19 2011
    - Mac OS X Snow Leopard support thanks to Svetoslav Agafonkin

0.06  Mon Oct 01 10:33:19 2011
    - Avoids segfaulting on attempting to open a locked DB, 
    - thanks to David Leadbeater

0.07  Mon Nov 14 02:33:19 2011
    - MS Windows support thanks to Julian Lishev <julian.lishev@imperia.net>

0.08 Thu Aug 14 04:49:19 2014
    - x86 support



Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
use ExtUtils::MakeMaker::Config;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

my %mk;
unless ($^O =~ /Win/i && $^O !~ /darwin/i) {
	system("sh ./build_detect_platform");
	open(MK,"build_config.mk") or die;
	map { $mk{$1} = $2 if(/^(.+?)\=(.*)$/) } <MK>;
	close(MK);
}

# fix: why MM_Unix->init_dirscan() haven't found these O_FILES?
my $OBJECT = "LevelDB.obj db/builder.obj db/db_impl.obj db/db_iter.obj db/filename.obj db/dbformat.obj db/log_reader.obj db/log_writer.obj db/memtable.obj db/repair.obj db/table_cache.obj db/version_edit.obj db/version_set.obj db/write_batch.obj table/block.obj table/block_builder.obj table/format.obj table/iterator.obj table/merger.obj table/table.obj table/table_builder.obj table/two_level_iterator.obj util/arena.obj util/cache.obj util/coding.obj util/comparator.obj util/crc32c.obj util/env.obj util/hash.obj util/histogram.obj util/logging.obj util/options.obj util/status.obj";


if ($^O =~ /Win/i && $^O !~ /darwin/i) {
	*MY::dynamic_lib = sub {
		package MY;
		my $lib_rules = shift->SUPER::dynamic_lib(@_);
		
		# Add deploy functionality for the extra dlls.
		$lib_rules .= <<END;
\t\$(NOECHO) \$(MOD_INSTALL) deploy blib
END

		$lib_rules;
	};
	
	$OBJECT .= " port/port_win.obj util/env_boost.obj util/win_logger.obj";
	
	WriteMakefile(
		NAME              	=> 'Tie::LevelDB',
		VERSION_FROM      	=> 'lib/Tie/LevelDB.pm', # finds $VERSION
		PREREQ_PM         	=> {}, # e.g., Module::Name => 1.1
		($] >= 5.005 ?     ## Add these new keywords supported since 5.005
		  (ABSTRACT_FROM  	=> 'lib/Tie/LevelDB.pm', # retrieve abstract from module
		   AUTHOR         	=> 'martin <martin@>') : ()),
		LIBS              	=> [''],
		DEFINE            	=> ("-D_WINDOWS -D_USRDLL -DLEVELDB_EXPORTS -DLEVELDB_PLATFORM_WINDOWS -DOS_WIN -D_CRT_SECURE_NO_WARNINGS " . '-Fo$@'),
		INC               	=> '-I. -Iinclude -Iport -Iutil -Iboost_1_47',
		OBJECT            	=> $OBJECT,
		CC				    => 'cl',
		CCFLAGS				=> $Config{ccflags} . ' -TP -EHsc',
		LDDLFLAGS			=> $Config{lddlflags} . ' -libpath:"boost_1_47\lib" boost_filesystem-vc100-mt-1_47.lib boost_date_time-vc100-mt-1_47.lib boost_thread-vc100-mt-1_47.lib boost_system-vc100-mt-1_47.lib msvcprt.lib',
		XSOPT       	    => '-C++',
		TYPEMAPS	        => ['perlobject.map'],
	);
} else {
	$OBJECT .= " port/port_posix.obj util/env_posix.obj";
	WriteMakefile(
		NAME              => 'Tie::LevelDB',
		VERSION_FROM      => 'lib/Tie/LevelDB.pm', # finds $VERSION
		PREREQ_PM         => {}, # e.g., Module::Name => 1.1
		($] >= 5.005 ?     ## Add these new keywords supported since 5.005
		  (ABSTRACT_FROM  => 'lib/Tie/LevelDB.pm', # retrieve abstract from module
		   AUTHOR         => 'martin <martin@>') : ()),
		LIBS              => [$mk{SNAPPY} ? '-lsnappy':''],
		DEFINE            => ($mk{SNAPPY} ? '-DSNAPPY':'').
								" ".$mk{PLATFORM_CFLAGS}." ".$mk{PORT_CFLAGS}
								.' -o $@',
		INC               => '-I. -Iinclude -Iport -Iutil',
		#OBJECT            => '$(O_FILES)', # link all the C files too
		OBJECT            => $OBJECT,

		'CC'			    => 'c++ '.$mk{PLATFORM_CFLAGS}." ".$mk{PORT_CFLAGS},
		'LD'			    => 'c++ '.$mk{PLATFORM_LDFLAGS},
		'XSOPT'             => '-C++',
		'TYPEMAPS'          => ['perlobject.map' ],
	);
}


lib/Tie/LevelDB.pm
package Tie::LevelDB;

use 5.010001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Tie::LevelDB ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.08';

require XSLoader;
XSLoader::load('Tie::LevelDB', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Tie::LevelDB - A Perl Interface to the Google LevelDB NoSQL database

=head1 SYNOPSIS

  use Tie::LevelDB;

  tie my %hash, 'Tie::LevelDB', "/tmp/testdb";
  # Use the %hash array
  untie %hash;

  -- OR --

  use Tie::LevelDB; 

  my $db = new Tie::LevelDB::DB("/tmp/testdb");
  $db->Put("Google","Don't be evil!");
  print $db->Get("Google")."\n";
  $db->Delete("Google");

  my $batch = new Tie::LevelDB::WriteBatch;
  $batch->Delete("Google");
  $batch->Put("Microsoft","Where Do you Want to Go Today?");
  $db->Write($batch);

  my $it = $db->NewIterator;
  for($it->SeekToFirst;$it->Valid;$it->Next) {
     print $it->key.": ".$it->value."\n";
  }

=head1 DESCRIPTION

B<Tie::LevelDB> is the Perl Interface for Google NoSQL database called
I<LevelDB>. See L<http://code.google.com/p/leveldb/> for more details.

Interface is implemented both as a reflection of an original LevelDB 
C++ API and a Perl-ish TIEHASH mechanism.

=head2 EXPORT

None by default.

=head2 LIMITATIONS

LevelDB does not support storing of C<undef> values. 
If C<undef> is stored, the key is C<deleted> instead.

Perl support for Options specification is not covered.

To use SNAPPY compression method, install it from 
L<http://code.google.com/p/snappy> first and then re-install
this module.

LevelDB sources (version 2011-07-29) are bundled with this packages. 

=head1 SEE ALSO

L<http://code.google.com/p/leveldb/>, L<DB_File(3)>, L<tie>.

=head1 AUTHOR

Martin Sarfy, E<lt>martin.sarfy@sokordia.czE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011 by Sokordia, s.r.o., L<http://www.sokordia.cz>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.


=cut
